### PR TITLE
FreeBSD module --enable-debug --enable-invariants

### PR DESCRIPTION
**Pull request by freqlabs**
### Motivation and Context

The FreeBSD module build infrastructure currently is unaware of the --enable-debug flag to ./configure, requiring WITH_DEBUG=true in the env or on the command line instead. Similarly, WITH_INVARIANTS=true is required when building for a kernel with the INVARIANTS option in KERNCONF. These idiosyncrasies are a trap for developers unaware of their existence.

### Description

Wire up the --enable-debug flag for configure to the FreeBSD module
build.  Add --enable-invariants.

The running FreeBSD kernel config is used to detect whether to enable
INVARIANTS if not explicitly specified with --enable-invariants or
--disable-invariants.

### How Has This Been Tested?

Configured with --enable-debug, ran gmake --trace -C module and observed the flags chosen by module/Makefile.bsd, then gmake --trace -C module WITH_INVARIANTS=true to confirm that the command line (or environment) take precedence over the configure flags (or lack thereof).

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: config/zfs-build.m4 [hunk 1]
```diff
@@ -152,6 +155,28 @@ AC_DEFUN([ZFS_AC_DEBUG_KMEM_TRACKING], [
 	AC_MSG_RESULT([$enable_debug_kmem_tracking])
 ])
 
+dnl #
+dnl # Disabled by default, enables INVARIANTS features in the FreeBSD kernel
+dnl # module.  This feature must be used when building for a FreeBSD kernel
+dnl # with "options INVARIANTS" in the KERNCONF and must not be used when the
+dnl # INVARIANTS option is absent.
+dnl #
```
Comment (behlendorf) [inline config/zfs-build.m4:163]: Is this something we can dynamically check for by default on FreeBSD so it always does the right thing?  Unless it's explicitly overridden by the config option of course.

**Review by behlendorf — COMMENTED** [review]
Great, keeping this behavior consistent will I'm sure avoid a lot of confusion.

File: config/zfs-build.m4 (hunk 1 above)
Comment (freqlabs) [reply config/zfs-build.m4:163]: Yes, I can grep for it in a sysctl that gives the running kernel config. I just need to think a bit on the m4.

Comment (behlendorf) [reply config/zfs-build.m4:163]: The `ZFS_AC_CONFIG_USER_*` m4 macros provides some simple examples of the check behavior you could pretty easily adapt.

**Comment by freqlabs** [comment]
* Detect INVARIANTS support in running kernel config by default
This changes the default behavior in a way that may impact cross-building the module for a different FreeBSD kernel than is currently running. Fortunately, the common situation this occurs in is when building the FreeBSD port for openzfs-kmod, which doesn't use the autoconf/autotools/automake parts anyway. The kmod port simply builds the module with Makefile.bsd and the environment configured through the port options.

**Review by behlendorf — APPROVED** [review]

**Pull request merged**